From: Chris Wright <chrisw@redhat.com>

    - src/qemu/qemu_driver.c: Update qemuGetPciHostDeviceList to take a
      hostdev list and count directly, rather than getting this indirectly
      from domain definition. This will allow reuse for the attach-device case.

diff -urp libvirt-0.8.1-p1/src/qemu/qemu_driver.c libvirt-0.8.1-p2/src/qemu/qemu_driver.c
--- libvirt-0.8.1-p1/src/qemu/qemu_driver.c	2010-07-21 08:40:51.272959807 -0400
+++ libvirt-0.8.1-p2/src/qemu/qemu_driver.c	2010-07-27 12:00:16.152760690 -0400
@@ -3189,7 +3189,7 @@ cleanup:
 
 
 static pciDeviceList *
-qemuGetPciHostDeviceList(virDomainDefPtr def)
+qemuGetPciHostDeviceList(virDomainHostdevDefPtr *hostdevs, int nhostdevs)
 {
     pciDeviceList *list;
     int i;
@@ -3197,8 +3197,8 @@ qemuGetPciHostDeviceList(virDomainDefPtr
     if (!(list = pciDeviceListNew()))
         return NULL;
 
-    for (i = 0 ; i < def->nhostdevs ; i++) {
-        virDomainHostdevDefPtr hostdev = def->hostdevs[i];
+    for (i = 0 ; i < nhostdevs ; i++) {
+        virDomainHostdevDefPtr hostdev = hostdevs[i];
         pciDevice *dev;
 
         if (hostdev->mode != VIR_DOMAIN_HOSTDEV_MODE_SUBSYS)
@@ -3238,7 +3238,7 @@ qemuUpdateActivePciHostdevs(struct qemud
     if (!def->nhostdevs)
         return 0;
 
-    if (!(pcidevs = qemuGetPciHostDeviceList(def)))
+    if (!(pcidevs = qemuGetPciHostDeviceList(def->hostdevs, def->nhostdevs)))
         return -1;
 
     for (i = 0; i < pciDeviceListCount(pcidevs); i++) {
@@ -3266,7 +3266,7 @@ qemuPrepareHostPCIDevices(struct qemud_d
     int i;
     int ret = -1;
 
-    if (!(pcidevs = qemuGetPciHostDeviceList(def)))
+    if (!(pcidevs = qemuGetPciHostDeviceList(def->hostdevs, def->nhostdevs)))
         return -1;
 
     /* We have to use 3 loops here. *All* devices must
@@ -3418,7 +3418,7 @@ qemuDomainReAttachHostDevices(struct qem
     if (!def->nhostdevs)
         return;
 
-    if (!(pcidevs = qemuGetPciHostDeviceList(def))) {
+    if (!(pcidevs = qemuGetPciHostDeviceList(def->hostdevs, def->nhostdevs))) {
         virErrorPtr err = virGetLastError();
         VIR_ERROR(_("Failed to allocate pciDeviceList: %s"),
                   err ? err->message : "");

